### PR TITLE
Fix possible problems with ClusterClient Discovery

### PR DESCRIPTION
**Pull request by Arkatufus**
## Changes

* Limits the number of initial contacts being used in ClusterClient Discovery to improve lost connection detection
* Randomly choose the initial contacts to prevent thundering herd problem

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs [hunk 1]
```diff
@@ -49,6 +50,7 @@ private sealed record ResolveResult(Contact Contact, IActorRef? Subject);
     private readonly string _targetActorSystemName;
     private readonly string _receptionistName;
     private readonly string _transportProtocol;
+    private readonly int _numberOfContacts;
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs:54]: This is just the number of initial contacts we need, right?

File: src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs [hunk 2]
```diff
@@ -229,6 +236,40 @@ async Task<ResolveResult[]> VerifyContacts()
...
+        Shuffle(fullContact);
+        return fullContact.Take(count).ToArray();
+    }
+
+    /// <summary>
+    /// Fisher-Yates in-place array shuffle algorithm
+    /// </summary>
+    /// <param name="array"></param>
+    /// <typeparam name="T"></typeparam>
+    private static void Shuffle<T>(T[] array)
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: Do we have another implementation of this lying around somewhere? I could have sworn I'd written one before

**Review by Aaronontheweb — COMMENTED** [review]
Left some general comments

File: src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs (hunk 1 above)
Comment (Arkatufus) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs:54]: it limits the number of contacts we use when we instantiate the ClusterClient actor. Too many of these will set the initial contacts to a higher than optimum number. The thing is that ClusterClient initial contacts never gets trimmed when they can't be contacted, if we use the whole discovery resolve result as the ClusterClient initial contact list, there will be a big burst of network connection attempt when the ClusterClient actor starts up and whenever its internal contact list could not be contacted anymore.

Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs:54]: Got it, makes sense

File: src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs (hunk 2 above)
Comment (Arkatufus) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: no idea

Comment (Arkatufus) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: There is one in the Akka.Util.Internal.ArrayExtensions, but its internal

Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: I would add a link to that file and just use that (or you can add a friend assembly to Akka.Discovery; link is probably cleaner.)

Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: That's just a nitpick though

Comment (Arkatufus) [reply src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClientDiscovery.cs]: Done, turns out the core Akka project is already a friend of Akka.Cluster.Tools

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**